### PR TITLE
feat: SMS/Whatsapp to attendee for teams

### LINKED ISSUE #13897: [CAL-3430] [CAL-3233] Bring back SMS reminders for teams
[https://threads.com/thread/34562092433/34562108801?s=4w4qJ3A4yEg6BpFDxe6syY](https://threads.com/thread/34562092433/34562108801?s=4w4qJ3A4yEg6BpFDxe6syY)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3233](https://linear.app/calcom/issue/CAL-3233/bring-back-sms-reminders-for-teams)</sub>

<sub>[CAL-3430](https://linear.app/calcom/issue/CAL-3430/[cal-3233]-bring-back-sms-reminders-for-teams)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

Before SMS/Whatsapp to attendee was only available for Orgs, now it will be available for all teams. It adds rate limit to sending SMS + an Admin view (localhost:3000/settings/admin/lockedSMS) to lock users and teams to send SMS.

Fixes #13897

### CODE REVIEW

File: packages/features/ee/workflows/lib/reminders/providers/twilioProvider.ts [hunk 1]
```diff
@@ -90,3 +136,44 @@ export const verifyNumber = async (phoneNumber: string, code: string) => {
...
+      return true;
+    }
+
+    const user = await prisma.user.findFirst({
+      where: {
+        id: userId,
+      },
+    });
+    return user?.smsLockState === SMSLockState.LOCKED;
+  }
```
Comment (joeauyeung) [inline packages/features/ee/workflows/lib/reminders/providers/twilioProvider.ts:178]: NIT: Can we actually use an `OR` condition in the prisma query against the `userId` to see if the user is locked.

**Review by joeauyeung — DISMISSED** [review]
Everything looks good! Nothing has really changed code wise in terms of sending SMS/Whatsapp messages it was just disabling the feature. I have a few nits but nothing blocking.

File: packages/features/ee/workflows/lib/reminders/reminderScheduler.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ const processWorkflowStep = async (
 ) => {
   if (isSMSAction(step.action)) {
     await checkSMSRateLimit({
-      identifier: `sms:${workflow.userId ? "user:" : "team:"}${workflow.userId || workflow.teamId}`,
```
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/reminders/reminderScheduler.ts]: Also team workflows have a `userId` (the user that created the workflow). We need to check for `teamId` here

File: packages/lib/checkRateLimitAndThrowError.ts [hunk 3]
```diff
@@ -58,15 +58,18 @@ async function changeSMSLockState(identifier: string, status: SMSLockState) {
     await prisma.user.update({
       where: {
         id: userId,
+        profiles: { none: {} },
```
Comment (CarinaWolli) [inline packages/lib/checkRateLimitAndThrowError.ts:61]: @joeauyeung  don't lock user's that belong to an org

File: packages/lib/checkRateLimitAndThrowError.ts [hunk 4]
```diff
@@ -58,15 +58,18 @@ async function changeSMSLockState(identifier: string, status: SMSLockState) {
...
       data: {
         smsLockState: status,
       },
     });
   } else {
-    await prisma.user.update({
+    await prisma.team.update({
       where: {
         id: teamId,
+        parentId: null,
```
Comment (CarinaWolli) [inline packages/lib/checkRateLimitAndThrowError.ts:71]: @joeauyeung  don't lock orgs or org teams

File: packages/features/ee/workflows/lib/reminders/reminderScheduler.ts [hunk 5]
```diff
@@ -48,7 +48,7 @@ const processWorkflowStep = async (
...
+      identifier: `sms:${workflow.teamId ? "team:" : "user:"}${workflow.teamId || workflow.userId}`,
```
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/reminders/reminderScheduler.ts:51]: Also team workflows have a `userId` (the user that created the workflow). We need to check for `teamId` here

File: apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        <tbody className="divide-subtle divide-y rounded-md">
+          {users.map((user) => (
+            <Row key={`user-${user.id}`}>
+              <Cell widthClassNames="w-auto">
+                <div className="min-h-10 flex items-center">
+                  {" "}
+                  <Avatar
+                    size="md"
+                    alt={`Avatar of ${user.username || "Nameless"}`}
+                    imageSrc={`${WEBAPP_URL}/${user.username ? user.username : "default"}/avatar.png`}
```
Comment (Udit-takkar) [inline apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx:104]: we should use `getUserAvatarUrl` here

**Review by Udit-takkar — COMMENTED** [review]
can you also block orgs teams?

Getting this error

Screenshot 2024-04-22 at 8 20 24 PM

File: apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            </Row>
+          ))}
+          {teams.map((team) => (
+            <Row key={`team-${team.id}`}>
+              <Cell widthClassNames="w-auto">
+                <div className="min-h-10 flex items-center">
+                  <Avatar
+                    size="md"
+                    alt={`Avatar of ${team.name}`}
+                    imageSrc={`${WEBAPP_URL}/${team.slug ? team.slug : "default"}/avatar.png`}
```
Comment (Udit-takkar) [inline apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx:128]: here we should probably use  const profileImageSrc = getPlaceholderAvatar(team.logoUrl || team.parent?.logoUrl, team.name);

**Comment by CarinaWolli** [comment]
> can you also block orgs teams?

orgs can't be blocked. also not their users and teams

File: apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+type Team = {
+  id: number;
+  name: string;
+  smsLockState: SMSLockState;
+  slug: string | null;
+};
+
+function UsersTable({ setSMSLockState }: Props) {
+  const { data: usersAndTeams } = trpc.viewer.admin.getSMSLockStateTeamsUsers.useQuery();
```
Comment (Udit-takkar) [inline apps/web/pages/settings/admin/lockedSMS/UsersTable.tsx:29]: Not urgent: use infinite query here. we can probably use it later

**Review by Udit-takkar — COMMENTED** [review]
LGTM. works fine for me. will wait for @joeauyeung re review before merging

**Review by joeauyeung — APPROVED** [review]
No big changes since my last approval. Approving again.

**Comment by keithwillcode** [comment]
@CarinaWolli @joeauyeung @Udit-takkar why was this PR merged with 0 tests added to verify SMS locking logic?

**Pull request merged**